### PR TITLE
Discussion for  possible changes for APL package

### PR DESCRIPTION
**Pull request by ClemensKohl**
This is a large pull request with all the changes made on my own repository. They have been tested and as far as I can tell all tests including BiocCheck are passed. 
As the changes are substantial, I would appreciate if you @elagralinska could review them (there is a very helpful Review changes function).

Some changes are not necessary or need more testing (e.g. random direction method for apl_score). We can hold off merging into main until all these things are done. Changes deemed not necessary (please mark the parts in the code that you find unhelpful) I can remove before merging into dev.

Any other suggestions/opinions for changes we maybe can also discuss here.

### CODE REVIEW

File: R/CA.R [hunk 1]
```diff
@@ -14,43 +15,220 @@ NULL
...
+#' @param theta Overdispersion parameter. By default set to 100 as described in
+#' Lause and Berens, 2021 (see references).
+#' 
+#' @references 
+#' Lause, J., Berens, P. & Kobak, D. Analytic Pearson residuals for 
+#' normalization of single-cell RNA-seq UMI data. Genome Biol 22, 258 (2021). 
+#' https://doi.org/10.1186/s13059-021-02451-7
+#' 
+#' @inherit calc_residuals return
+comp_NB_residuals <- function(mat,
```
Comment (elagralinska) [inline R/CA.R:80]: Is it needed for Cabinet?

File: man/subset_dims.Rd
Comment (elagralinska) [inline man/subset_dims.Rd:1]: ?

File: man/var_rows.Rd
Comment (elagralinska) [inline man/var_rows.Rd:1]: Are freeman-tukey and NB residuals needed?

File: man/recompute.Rd
Comment (elagralinska) [inline man/recompute.Rd:1]: Not sure what is it about, let's discuss it together.

File: man/random_direction_cutoff.Rd
Comment (elagralinska) [inline man/random_direction_cutoff.Rd:1]: I would get rid of this change.

File: R/convert.R [hunk 2]
```diff
@@ -15,15 +15,24 @@ NULL
...
-  if(isTRUE(rm_zeros)){
+  stopifnot(is(mat, "matrix") | is(mat, "Matrix"))
+  
+  if(is.null(calist$params)){
+      warning("No parameters provided for recalculation!")
+      calist$params <- list()
+      
+  }
+  
+  # if (is.null(calist$top_rows)) top <- nrow(mat)
```
Comment (elagralinska) [inline R/convert.R:30]: Difficult to follow this part, let's talk about it.

File: R/convert.R [hunk 3]
```diff
@@ -356,7 +385,9 @@ setMethod(f = "as.cacomp",
 
   ca_list <- list("std_coords_cols" = Seurat::Embeddings(obj, reduction = "CA"),
                   "D" = Seurat::Stdev(obj, reduction = "CA"),
-                  "prin_coords_rows" = Seurat::Loadings(obj, reduction = "CA"))
+                  "prin_coords_rows" = Seurat::Loadings(obj, reduction = "CA"),
+                  "params" = obj@reductions$CA@misc)
```
Comment (elagralinska) [inline R/convert.R:389]: Let's talk about params.

File: man/apl_score.Rd [hunk 4]
```diff
@@ -41,6 +42,9 @@ decomposition from the python package torch.}
 object.
 This implementation dramatically speeds up computation compared to `svd()` 
 in R.}
+
```
Comment (elagralinska) [inline man/apl_score.Rd:45]: If we get rid of "random", this also needs to be removed.

File: man/apl_score.Rd [hunk 5]
```diff
@@ -55,8 +59,14 @@ with the condition and how associated it is with other conditions.
 The score is calculated by permuting the values of each row to determine the 
 cutoff angle of the 99% quantile.
 \deqn{S_{alpha}(x,y)=x-\frac{y}{\tan\alpha}}
-By default the permutation is repeated 10 times, but for very large matrices 
+By default the permutation is repeated 10 times (for random direction min.
```
Comment (elagralinska) [inline man/apl_score.Rd:62]: Where does 300 come from? Any tests?

File: man/apl_score.Rd [hunk 6]
```diff
@@ -55,8 +59,14 @@ with the condition and how associated it is with other conditions.
 The score is calculated by permuting the values of each row to determine the 
 cutoff angle of the 99% quantile.
 \deqn{S_{alpha}(x,y)=x-\frac{y}{\tan\alpha}}
-By default the permutation is repeated 10 times, but for very large matrices 
+By default the permutation is repeated 10 times (for random direction min. 
+300 repetition is recommended!), but for very large matrices 
 this can be reduced.
+The method "permutation" permutes the columns in each row and calculates
```
Comment (elagralinska) [inline man/apl_score.Rd:65]: Same as above.

File: man/run_cacomp.Rd [hunk 7]
```diff
@@ -23,29 +26,39 @@ samples/cells in columns.
...
 
 \item{coords}{Logical. Indicates whether CA standard coordinates should be 
-calculated. Default TRUE}
+calculated.}
 
 \item{princ_coords}{Integer. Number indicating whether principal 
 coordinates should be calculated for the rows (=1), columns (=2), 
-both (=3) or none (=0).
-Default 1.}
+both (=3) or none (=0).}
```
Comment (elagralinska) [inline man/run_cacomp.Rd:33]: (is anyone using it?)

File: man/run_cacomp.Rd [hunk 8]
```diff
@@ -23,29 +26,39 @@ samples/cells in columns.
...
 
 \item{python}{A logical value indicating whether to use singular-value 
 decomposition from the python package torch.
 This implementation dramatically speeds up computation compared to `svd()` 
-in R.}
```
Comment (elagralinska) [inline man/run_cacomp.Rd:36]: "Full SVD" - rename?

File: man/run_cacomp.Rd [hunk 9]
```diff
@@ -23,29 +26,39 @@ samples/cells in columns.
...
 
-\item{inertia}{Logical.. Whether total, row and column inertias should be 
-calculated and returned. Default TRUE.}
+\item{inertia}{Logical. Whether total, row and column inertias should be 
+calculated and returned.}
 
 \item{rm_zeros}{Logical. Whether rows & cols containing only 0s should be 
-removed. Keeping zero only rows/cols might lead to unexpected results. 
-Default TRUE.}
+removed. Keeping zero only rows/cols might lead to unexpected results.}
```
Comment (elagralinska) [inline man/run_cacomp.Rd:51]: Same comment about types of residuals

File: man/runAPL.Rd [hunk 10]
```diff
@@ -116,24 +124,30 @@ sequencing a count matrix, gene expression values with genes in rows and
...
+\item{top}{Integer. Number of most variable rows to retain. 
+Set NULL to keep all.}
+
+\item{clip}{logical. Whether residuals should be clipped if they are 
+higher/lower than a specified cutoff}
 
 \item{score}{Logical. Whether rows should be scored and ranked. Ignored when 
 a vector is supplied to mark_rows. Default TRUE.}
 
+\item{score_method}{Method to calculate the cutoff. Either "random" for random
```
Comment (elagralinska) [inline man/runAPL.Rd:148]: To be removed in case we get rid of the randomized method

File: man/runAPL.Rd [hunk 11]
```diff
@@ -142,26 +156,27 @@ in the plot.}
 
 \item{reps}{Integer. Number of permutations during scoring. Default 3.}
 
-\item{python}{A logical value indicating whether to use singular value 
+\item{python}{A logical value indicating whether to use singular-value 
 decomposition from the python package torch.
 This implementation dramatically speeds up computation compared to `svd()` 
-in R.}
+in R when calculating the full SVD. This parameter only works when dims==NULL
```
Comment (elagralinska) [inline man/runAPL.Rd:162]: "Full SVD" - rename?

File: man/runAPL.Rd [hunk 12]
```diff
@@ -142,26 +156,27 @@ in the plot.}
...
 should be labeled with text. Default TRUE.}
 
 \item{col_labs}{Logical. Whether labels for columns indicated by cols_idx 
-should be labeled with text. Default TRUE.}
+shouls be labeled with text. Default FALSE.}
 
-\item{type}{"ggplot"/"plotly". For a static plot a string "ggplot", for an 
-interactive plot "plotly". Default "plotly".}
+\item{type}{"ggplot"/"plotly". For a static plot a string "ggplot", 
+for an interactive plot "plotly". Default "ggplot".}
```
Comment (elagralinska) [inline man/runAPL.Rd:172]: Plotly still in?

File: man/permutation_cutoff.Rd
Comment (elagralinska) [inline man/permutation_cutoff.Rd:1]: Does it refer to randomized directions? If yes, remove it?

File: man/comp_ft_residuals.Rd
Comment (elagralinska) [inline man/comp_ft_residuals.Rd:1]: Should we get rid of Freeman-Tukey residuals?

File: man/comp_NB_residuals.Rd
Comment (elagralinska) [inline man/comp_NB_residuals.Rd:1]: Same comment about residuals as before

File: man/clip_residuals.Rd
Comment (elagralinska) [inline man/clip_residuals.Rd:1]: Same comment about residuals

File: man/calc_residuals.Rd
Comment (elagralinska) [inline man/calc_residuals.Rd:1]: Same comment about residuals

File: man/cacomp.Rd [hunk 13]
```diff
@@ -67,30 +95,39 @@ and samples/cells in columns.
...
 coordinates should be calculated for the rows (=1), columns (=2), 
-both (=3) or none (=0).
-Default 1.}
+both (=3) or none (=0).}
 
 \item{python}{A logical value indicating whether to use singular-value 
 decomposition from the python package torch.
 This implementation dramatically speeds up computation compared to `svd()` 
-in R.}
+in R when calculating the full SVD. This parameter only works when dims==NULL
```
Comment (elagralinska) [inline man/cacomp.Rd:107]: "Full SVD" - rename it?

File: man/cacomp.Rd [hunk 14]
```diff
@@ -67,30 +95,39 @@ and samples/cells in columns.
...
-\item{inertia}{Logical.. Whether total, row and column inertias should be 
-calculated and returned. Default TRUE.}
+\item{inertia}{Logical. Whether total, row and column inertias should be 
+calculated and returned.}
 
 \item{rm_zeros}{Logical. Whether rows & cols containing only 0s should be 
-removed.
-Keeping zero only rows/cols might lead to unexpected results. Default TRUE.}
+removed. Keeping zero only rows/cols might lead to unexpected results.}
+
```
Comment (elagralinska) [inline man/cacomp.Rd:121]: Same comment about residuals

File: R/plot.R [hunk 15]
```diff
@@ -1,6 +1,11 @@
 #' @include constructor.R
 NULL
```
Comment (elagralinska) [inline R/plot.R:3]: ToDos!

File: R/apl.R [hunk 16]
```diff
@@ -137,8 +138,14 @@ apl_coords <- function(caobj, group, calc_rows = TRUE, calc_cols = TRUE){
 #' The score is calculated by permuting the values of each row to determine the 
 #' cutoff angle of the 99% quantile.
 #' \deqn{S_{alpha}(x,y)=x-\frac{y}{\tan\alpha}}
-#' By default the permutation is repeated 10 times, but for very large matrices
```
Comment (elagralinska) [inline R/apl.R:140]: Random direction - get rid of it?

File: R/apl.R [hunk 17]
```diff
@@ -190,122 +199,248 @@ apl_coords <- function(caobj, group, calc_rows = TRUE, calc_cols = TRUE){
...
+  
   if (is(group, "character")){
     idx <- match(group, names)
     idx <- na.omit(idx)
     group <- idx
   }
-
-  if (caobj@dims == 1 && !is.empty(caobj@dims)){
-    row_num <- 1
+
```
Comment (elagralinska) [inline R/apl.R:229]: Get rid of it?

File: R/apl.R [hunk 18]
```diff
@@ -1031,7 +1151,9 @@ setGeneric("runAPL", function(obj,
                               dims = NULL,
                               nrow = 10,
                               top = 5000,
+                              clip = FALSE,
                               score = TRUE,
+                              score_method = "permutation",
```
Comment (elagralinska) [inline R/apl.R:1156]: To discuss.

**Review by elagralinska — COMMENTED** [review]
Made some comments + a few things to discuss before accepting/declining

**Comment by ClemensKohl** [comment]
Hi! I made the changes as discussed! Let me know if anything is missing. There is also now a copy of the main branch for safekeeping.

**Comment by elagralinska** [comment]
Hi Clemens, thank you for your work! Looks good to me, nothing is missing in my opinion.

**Review by elagralinska — APPROVED** [review]

**Pull request merged**